README.md
# Log

Logging is an important part of the application and having a consistent logging mechanism and structure is mandatory. With several teams writing different components that talk to each other, being able to read each others logs could be the difference between finding bugs quickly or wasting hours.

## License
This software is licensed under the Apache license. For full text see [LICENSE](./LICENSE)

## Examples
The following code provides some examples of logging.

### Logging Without Log Levels
```
func main() {
	var buf bytes.Buffer
	log.Init("LogExample", 0, log.DevWriter{Device: log.DevAll, Writer: &buf})

	Square(5)

	log.Shutdown()
	fmt.Printf(buf.String())
}

func Square(n int) int {
	context := "SquareTwice"
	function := "Square"

	log.Start(context, function)

	sq := n * n
	log.Warnf(context, function, "Square 1: %d\n", sq)

	sq = 0
	for i := 0; i < n; i++ {
		sq += n
	}
	log.Tracef(context, function, "Square 2: %d\n", sq)

	log.Complete(context, function)

	return sq
}
```
The output produced
```
2016/08/22 13:05:42.596: LogExample[29218]: main.go#24: SquareTwice: Square: Started:
2016/08/22 13:05:42.596: LogExample[29218]: main.go#27: SquareTwice: Square: Warning: Square 1: 25
2016/08/22 13:05:42.596: LogExample[29218]: main.go#33: SquareTwice: Square: Trace: Square 2: 25
2016/08/22 13:05:42.596: LogExample[29218]: main.go#35: SquareTwice: Square: Completed:
```

### Logging With Log Levels
```
var warnLogger = log.NewLogger("squareLogger", func() int { return log.LevelWarning })

func main() {
	var buf bytes.Buffer
	log.Init("LogExample", 0, log.DevWriter{Device: log.DevAll, Writer: &buf})

	Square(5)

	log.Shutdown()
	fmt.Printf(buf.String())
}

func Square(n int) int {
	context := "SquareTwice"
	function := "Square"

	log.Start(context, function)

	sq := n * n
	warnLogger.Warnf(context, function, "Square 1: %d\n", sq)

	sq = 0
	for i := 0; i < n; i++ {
		sq += n
	}
	// NOTE: The following TRACE line will not be emitted because the log level
	// of the `warnLogger` is at the `Warning` level.
	warnLogger.Tracef(context, function, "Square 2: %d\n", sq)

	log.Complete(context, function)

	return sq
}
```
The output produced
```
2016/08/22 13:20:41.594: LogExample[29406]: main.go#26: SquareTwice: Square: Started:
2016/08/22 13:20:41.594: LogExample[29406]: main.go#29: SquareTwice: Square: Warning: Square 1: 25
2016/08/22 13:20:41.594: LogExample[29406]: main.go#37: SquareTwice: Square: Completed:
```

## Project Info
Package docs: https://godoc.org/github.com/Comcast/go-log/log


CHANGELOG.md
## Changelog

### 11-21-2017

- [Issue #8](https://github.com/Comcast/go-log/issues/8) - Update timestamp nanosecond precision to 9 digits.

### 11-13-2017
- [Issue #5](https://github.com/Comcast/go-log/issues/5) - Exports SafeBuffer so that (external) tests can use it.

### 11-03-2017
- [Issue #3](https://github.com/Comcast/go-log/issues/3) - Emits logs in bulk for better performance
  - Introduced bulk logs timer (defaults to 1s)
  - Made stall timeout variable


log/log_examples_test.go
/**
* Copyright 2016 Comcast Cable Communications Management, LLC
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
* http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
 */

package log_test

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"testing"
	"time"

	"github.com/Comcast/go-log/log"
)

// ExampleStart provides a basic example for using the log package.
func ExampleStart() {
	// Init the log system using a buffer for testing.
	buf := new(log.SafeBuffer)
	log.InitTest("EXAMPLE", 10, log.DevWriter{Device: log.DevAll, Writer: buf})

	{
		log.Start("1234", "Basic")

		v, err := strconv.ParseInt("10", 10, 64)
		if err != nil {
			log.CompleteErr(err, "1234", "Basic")
			return
		}

		log.Completef("1234", "Basic", "Conv[%d]", v)
	}

	log.Shutdown()
	fmt.Println(buf.String())
	// Output:
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Basic: Started:
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Basic: Completed: Conv[10]
}

// ExampleErr provides an example of logging an error.
func ExampleErr() {
	// Init the log system using a buffer for testing.
	buf := new(log.SafeBuffer)
	log.InitTest("EXAMPLE", 10, log.DevWriter{Device: log.DevAll, Writer: buf})

	{
		log.Start("1234", "Error")

		v, err := strconv.ParseInt("1080980980980980980898908", 10, 64)
		if err != nil {
			log.CompleteErr(err, "1234", "Error")

			// Flush the output for testing.
			log.Shutdown()
			fmt.Println(buf.String())
			return
		}

		log.Completef("1234", "Error", "Conv[%d]", v)
	}

	log.Shutdown()
	fmt.Println(buf.String())
	// Output:
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Error: Started:
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Error: Completed ERROR: strconv.ParseInt: parsing "1080980980980980980898908": value out of range
}

// ExampleDataKV provides an example of logging K/V pair data.
func ExampleDataKV() {
	// Init the log system using a buffer for testing.
	buf := new(log.SafeBuffer)
	log.InitTest("EXAMPLE", 10, log.DevWriter{Device: log.DevAll, Writer: buf})

	{
		log.Start("1234", "Data_KV")

		log.DataKV("1234", "Data_KV", "Value 1", 1)
		log.DataKV("1234", "Data_KV", "Hex Value 2", 0x00000002)

		log.Complete("1234", "Data_KV")
	}

	log.Shutdown()
	fmt.Println(buf.String())
	// Output:
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Data_KV: Started:
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Data_KV: DATA: Value 1: 1
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Data_KV: DATA: Hex Value 2: 2
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Data_KV: Completed:
}

// ExampleDataBlock provides an example of logging a block of data.
func ExampleDataBlock() {
	// Init the log system using a buffer for testing.
	buf := new(log.SafeBuffer)
	log.InitTest("EXAMPLE", 10, log.DevWriter{Device: log.DevAll, Writer: buf})

	{
		log.Start("1234", "Data_Block")

		data := `Test Data with
2 lines`

		log.DataBlock("1234", "Data_Block", data)

		log.Complete("1234", "Data_Block")
	}

	log.Shutdown()
	fmt.Println(buf.String())
	// Output:
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Data_Block: Started:
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Data_Block: DATA:
	// 	Test Data with
	// 	2 lines
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Data_Block: Completed:
}

type Message []byte

// Format implements the Formatter interface to produce logging output
// for this slice of bytes.
func (m Message) Format() string {
	var buf log.SafeBuffer

	// Message Bytes:	(0x0000) EE 6E 11 00 00 00 3E EA DE 18 00 00 2D 00 00 00
	// 					(0x0010) 00 00 00 00 00 0A 16 C3 9B 00 00 00 00 00 00 00

	fmt.Fprintf(&buf, "Message Bytes:\t")

	rows := (len(m) / 16) + 1
	l := len(m)
	for row := 0; row < rows; row++ {
		var r []byte
		st := row * 16

		if row < (rows - 1) {
			r = m[st : st+16]
		} else {
			r = m[st : st+(l-st)]
		}

		if row > 0 {
			fmt.Fprintf(&buf, "\t\t\t")
		}

		fmt.Fprintf(&buf, "(0x%.4X)", st)

		for i := 0; i < len(r); i++ {
			fmt.Fprintf(&buf, " %.2X", r[i])
		}
		fmt.Fprintf(&buf, "\n")
	}

	return buf.String()
}

// ExampleDataTrace provides an example of logging from a fmt.Stringer.
func ExampleDataTrace() {
	// Init the log system using a buffer for testing.
	buf := new(log.SafeBuffer)
	log.InitTest("EXAMPLE", 10, log.DevWriter{Device: log.DevAll, Writer: buf})

	{
		log.Start("1234", "Data_String")

		b1 := []byte{0xEE, 0x6E, 0x11, 0x00, 0x00, 0x00, 0x3E, 0xEA, 0xDE, 0x18, 0x00, 0x00, 0x2D, 0x00, 0x00, 0x00, 0x3E, 0xEA, 0xDE}
		b2 := []byte{0xEE, 0x6E, 0x11, 0x00, 0x00, 0x00, 0x3E, 0xEA, 0xDE, 0x18, 0x00, 0x00, 0x2D, 0x00, 0x00, 0x00, 0x3E, 0xEA, 0xDE}

		log.DataTrace("1234", "Data_String", Message(b1), Message(b2))

		log.Complete("1234", "Data_String")
	}

	log.Shutdown()
	fmt.Println(buf.String())
	// Output:
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Data_String: Started:
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Data_String: DATA:
	// 	Message Bytes:	(0x0000) EE 6E 11 00 00 00 3E EA DE 18 00 00 2D 00 00 00
	// 				(0x0010) 3E EA DE
	// 	Message Bytes:	(0x0000) EE 6E 11 00 00 00 3E EA DE 18 00 00 2D 00 00 00
	// 				(0x0010) 3E EA DE
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Data_String: Completed:
}

// ExampleTracef provides an example of logging from a fmt.Stringer and also tests newline handling.
func ExampleTracef() {
	// Init the log system using a buffer for testing.
	buf := new(log.SafeBuffer)
	log.InitTest("EXAMPLE", 10, log.DevWriter{Device: log.DevAll, Writer: buf})

	{
		// Messages without a newline should have one added, and message
		// that have a newline should *not* have an extra one added.
		log.Tracef("1234", "Tracef", "%s: %s", "1234", "Basic")
		log.Tracef("1234", "Tracef", "%s: %s\n", "1234", "Basic")
		log.Tracef("1234", "Tracef", "%s: %s", "ABCD", "Basic")
		log.Tracef("1234", "Tracef", "%s: %s\n", "ABCD", "Basic")
	}

	log.Shutdown()
	fmt.Println(buf.String())
	// Output:
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Tracef: Trace: 1234: Basic
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Tracef: Trace: 1234: Basic
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Tracef: Trace: ABCD: Basic
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Tracef: Trace: ABCD: Basic
}

// ExampleUplevel_Start provides an example of using the level up functionality.
func ExampleUplevel_Start() {
	// The following code would generate a log message with line
	// number 13 (the line number where handleErr is called) and function name
	// "example.someFunc".
	//
	//    1: package example
	//    2:
	//    3: func handleErr(err error, context interface{}, function string) {
	//    4:        ...
	//    5:        // log with caller's file, line, and function name.
	//    6:        log.Up1.Err(err, context, function)
	//    7:        ...
	//    8: }
	//    9:
	//   10: func someFunc() {
	//   11:        ...
	//   12:        if err := doSomething(); err != nil {
	//   13:                handleErr(err, context, "")
	//   14:        }
	//   15:        ...
	//   16: }

	// Init the log system using a buffer for testing.
	buf := new(log.SafeBuffer)
	log.InitTest("EXAMPLE", 10, log.DevWriter{Device: log.DevAll, Writer: buf})

	{
		levelUp := func(context interface{}, function string) {
			log.Up1.Tracef(context, function, "Test")
		}

		log.Start("1234", "Basic")
		levelUp("1234", "Basic")
		log.Complete("1234", "Basic")
	}

	log.Shutdown()
	fmt.Println(buf.String())

	// Output:
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Basic: Started:
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Basic: Trace: Test
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Basic: Completed:
}

// Example_multipleInit tests that when Init is called back to back,
// the log flushes and takes the change.
func Example_multipleInit() {
	// Init the log system using a buffer for testing.
	buf := new(log.SafeBuffer)
	log.InitTest("EXAMPLE", 10, log.DevWriter{Device: log.DevAll, Writer: buf})
	log.Start("1234", "Basic")
	log.Complete("1234", "Basic")

	// Now init the log with a buffer to disable logging
	log.InitTest("EXAMPLE", 10, log.DevWriter{Device: log.DevAll, Writer: ioutil.Discard})
	log.Start("1234", "SHOULD NOT BE DISPLAYED")
	log.Complete("1234", "SHOULD NOT BE DISPLAYED")

	log.Shutdown()
	fmt.Println(buf.String())
	// Output:
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Basic: Started:
	// 2009/11/10 15:00:00.000000000: EXAMPLE[69910]: file.go#512: 1234: Basic: Completed:
}

func BenchmarkTracef(b *testing.B) {
	log.InitTest("BENCHMARK", 10, log.DevWriter{Device: log.DevAll, Writer: ioutil.Discard})
	for i := 0; i < b.N; i++ {
		log.Tracef("context", "function", "This is a test %d this is a test %d this is a test %d", i, i, i)
	}
}

// ExampleSplunk provides an example of logging a message in a splunk-able format.
func ExampleSplunk() {
	// Init the log system using a buffer for testing.
	buf := new(log.SafeBuffer)
	log.InitTest("TestSplunk", 10, log.DevWriter{Device: log.DevAll, Writer: buf})

	sl1 := log.SplunkValue{1, 2, 3, 4}           // slice of ints.
	sl2 := log.SplunkValue{"123.123", "123.124"} // slice of strings.
	sl3 := log.SplunkValue{6, "123.123"}         // slice of mixed values.

	m := []log.SplunkPair{
		{Key: "Key1", Value: "Value1"},
		{Key: "RequestTime", Value: time.Date(2019, time.November, 10, 15, 0, 0, 0, time.UTC).UTC().Format("2006/01/02 15:04:05.000000000")},
		{Key: "MAC", Value: "010203040506"},
		{Key: "ResponseCode", Value: 0},
		{Key: "Slice", Value: sl1},
		{Key: "name1", Value: sl2},
		{Key: "name2", Value: sl3},
	}
	log.Splunk(m...)

	log.Splunk(log.SplunkPair{Key: "SecondKey", Value: "SecondValue"},
		log.SplunkPair{Key: "RequestTime", Value: time.Date(2019, time.November, 10, 15, 0, 0, 0, time.UTC).UTC().Format("2006/01/02 15:04:05.000000000")},
		log.SplunkPair{Key: "MAC", Value: "010203040507"},
		log.SplunkPair{Key: "ResponseCode", Value: 0},
		log.SplunkPair{Key: "Slice", Value: sl1},
		log.SplunkPair{Key: "name1", Value: sl2},
		log.SplunkPair{Key: "name2", Value: sl3})

	log.Shutdown()
	fmt.Println(buf.String())

	// Output:
	// 2009/11/10 15:00:00.000000000: Key1=Value1 RequestTime="2019/11/10 15:00:00.000000000" MAC=010203040506 ResponseCode=0 Slice=[1, 2, 3, 4] name1=[123.123, 123.124] name2=[6, 123.123]
	// 2009/11/10 15:00:00.000000000: SecondKey=SecondValue RequestTime="2019/11/10 15:00:00.000000000" MAC=010203040507 ResponseCode=0 Slice=[1, 2, 3, 4] name1=[123.123, 123.124] name2=[6, 123.123]
}


log/log_whitebox_test.go
/**
* Copyright 2016 Comcast Cable Communications Management, LLC
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
* http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
 */

package log

import (
	"bytes"
	"os"
	"regexp"
	"testing"
	"time"
)

func TestDtFile(t *testing.T) {
	const calldepth = 20
	t.Log("Given the need to get date time and file.")
	{
		expectedFuncName := "TestDtFile"
		dateTime, file, funcName, pid := dtFile(2, expectedFuncName)

		// At time of writing this function will return "testing.go#485". But adding
		// test might change the line number may change the second part and there's
		// no guarantee that go will always run tests inside testing.go. So we use a
		// regex to verify as close as possible.
		match, err := regexp.Match("[a-zA-Z0-9].go#\\d+", ([]byte)(file))
		if err != nil {
			t.Error("the regex is broken, please fix the test.", err)
			return
		}
		if !match {
			t.Error("\tfile should match with expected regex. ", failed)
		} else {
			t.Log("\tfile should match with expected regex. ", succeed)
		}

		// verify the process id
		if pid != os.Getpid() {
			t.Error("\tProcess ID should match.", failed)
		} else {
			t.Log("\tProcess ID should match.", succeed)
		}

		//verify funcname
		if funcName != expectedFuncName {
			t.Error("\tfuncName should match.", failed)
		} else {
			t.Log("\tfuncName should match.", succeed)
		}

		// verify that dateTime is roughly now. It's the best we can do.
		dt, err := time.Parse(layout, dateTime)
		if err != nil {
			t.Error("\tdateTime should be parsable.", failed)
		} else {
			t.Log("\tdateTime should be parsable.", succeed)
		}

		// If they're more then 5 seconds apart from each other then we either
		// have a performance problem on go-log, something terrible happened on
		// the machine performing the test, or dtFile is just plain wrong.
		timeDiff := time.Now().Sub(dt)
		if timeDiff > 5*time.Second {
			t.Error("\tNow-dateTime shoud be less then 5 second away.", failed)
		} else {
			t.Log("\tNow-dateTime shoud be less then 5 second away.", succeed)
		}
	}
	t.Log("Given a way too big caller depth")
	{
		expectedFuncName := "TestDtFile"
		dateTime, file, funcName, pid := dtFile(calldepth, expectedFuncName)

		// with a broken caller depth the filename returned should be unknown
		// and line number is zero
		if file != "unknown.go#0:" {
			t.Error("\tfile should match \"unknown.go#0:\". ", failed)
		} else {
			t.Log("\tfile should match \"unknown.go#0:\". ", succeed)
		}

		// verify the process id
		if pid != os.Getpid() {
			t.Error("\tProcess ID should match.", failed)
		} else {
			t.Log("\tProcess ID should match.", succeed)
		}

		// verify funcname, it should be unknown because the caller depth is too
		// big
		if funcName != "missing" {
			t.Error("\tfuncName should match.", failed)
		} else {
			t.Log("\tfuncName should match.", succeed)
		}

		// verify that dateTime is roughly now. It's the best we can do.
		dt, err := time.Parse(layout, dateTime)
		if err != nil {
			t.Error("\tdateTime should be parsable.", failed)
		} else {
			t.Log("\tdateTime should be parsable.", succeed)
		}

		// If they're more then 5 seconds apart from each other then we either
		// have a performance problem on go-log, something terrible happened on
		// the machine performing the test, or dtFile is just plain wrong.
		timeDiff := time.Now().Sub(dt)
		if timeDiff > 5*time.Second {
			t.Error("\tNow-dateTime shoud be less then 5 second away.", failed)
		} else {
			t.Log("\tNow-dateTime shoud be less then 5 second away.", succeed)
		}
	}

	t.Log("Given no funcName.")
	{
		// this is the actual function that is using this test.
		expectedFuncName := "testing.tRunner"
		dateTime, file, funcName, pid := dtFile(2, "")

		// At time of writing this function will return "testing.go#485". But adding
		// test might change the line number may change the second part and there's
		// no guarantee that go will always run tests inside testing.go. So we use a
		// regex to verify as close as possible.
		match, err := regexp.Match("[a-zA-Z0-9].go#\\d+", ([]byte)(file))
		if err != nil {
			t.Error("the regex is broken, please fix the test.", err)
			return
		}
		if !match {
			t.Error("\tfile should match with expected regex. ", failed)
		} else {
			t.Log("\tfile should match with expected regex. ", succeed)
		}

		// verify the process id
		if pid != os.Getpid() {
			t.Error("\tProcess ID should match.", failed)
		} else {
			t.Log("\tProcess ID should match.", succeed)
		}

		//verify funcname
		if funcName != expectedFuncName {
			t.Error("\tfuncName should match.", failed)
		} else {
			t.Log("\tfuncName should match.", succeed)
		}

		// verify that dateTime is roughly now. It's the best we can do.
		dt, err := time.Parse(layout, dateTime)
		if err != nil {
			t.Error("\tdateTime should be parsable.", failed)
		} else {
			t.Log("\tdateTime should be parsable.", succeed)
		}

		// If they're more then 5 seconds apart from each other then we either
		// have a performance problem on go-log, something terrible happened on
		// the machine performing the test, or dtFile is just plain wrong.
		timeDiff := time.Now().Sub(dt)
		if timeDiff > 5*time.Second {
			t.Error("\tNow-dateTime shoud be less then 5 second away.", failed)
		} else {
			t.Log("\tNow-dateTime shoud be less then 5 second away.", succeed)
		}
	}

	t.Log("Given a way too big caller depth and no funcName")
	{
		// here we are testing wether "missing" takes precedence over not giving
		// a funcName.
		expectedFuncName := "missing"
		dateTime, file, funcName, pid := dtFile(calldepth, "")

		// with a broken caller depth the filename returned should be unknown
		// and line number is zero
		if file != "unknown.go#0:" {
			t.Error("\tfile should match \"unknown.go#0:\". ", failed)
		} else {
			t.Log("\tfile should match \"unknown.go#0:\". ", succeed)
		}

		// verify the process id
		if pid != os.Getpid() {
			t.Error("\tProcess ID should match.", failed)
		} else {
			t.Log("\tProcess ID should match.", succeed)
		}

		// verify funcname, it should be unknown because the caller depth is too
		// big
		if funcName != expectedFuncName {
			t.Error("\tfuncName should match.", failed)
		} else {
			t.Log("\tfuncName should match.", succeed)
		}

		// verify that dateTime is roughly now. It's the best we can do.
		dt, err := time.Parse(layout, dateTime)
		if err != nil {
			t.Error("\tdateTime should be parsable.", failed)
		} else {
			t.Log("\tdateTime should be parsable.", succeed)
		}

		// If they're more then 5 seconds apart from each other then we either
		// have a performance problem on go-log, something terrible happened on
		// the machine performing the test, or dtFile is just plain wrong.
		timeDiff := time.Now().Sub(dt)
		if timeDiff > 5*time.Second {
			t.Error("\tNow-dateTime shoud be less then 5 second away.", failed)
		} else {
			t.Log("\tNow-dateTime shoud be less then 5 second away.", succeed)
		}
	}
}

func TestOutput(t *testing.T) {
	t.Log("Given no format passed to output.")
	{
		var buf SafeBuffer
		Init("TEST", 0, DevWriter{
			Device: DevAll,
			Writer: &buf,
		})

		// We expect this to generate some message because format is empty.
		output(&buf, "")

		// don't defer the shutdown because we need a clean start for the next
		// part of the test.
		Shutdown()

		if buf.String() != emptyMessage {
			t.Error("\tempty format should generate error message.", failed)
		}
		t.Log("\tempty format should generate error message.", succeed)

	}
	t.Log("Given no format passed to output.")
	{
		var buf bytes.Buffer
		Init("TEST", 0, DevWriter{
			Device: DevAll,
			Writer: &buf,
		})

		// Shutdown right now to trigger the no write on `output`.
		Shutdown()

		// We expect this to generate some message because format is empty.
		output(&buf, "")

		if buf.String() != "" {
			t.Error("\tempty format should contain nothing.", failed)
		}
		t.Log("\tempty format should contain nothing.", succeed)
	}
}

// This test doesn't actually work because if you pass nil to output another
// goroutine panics, not this one. I need a second opinion on this. However, I
// have added a check for nil in output to prevent it from panicking.
func TestOutputNilWriter(t *testing.T) {
	t.Log("Given no writer passed to output.")
	{
		var buf bytes.Buffer
		Init("TEST", 0, DevWriter{
			Device: DevAll,
			Writer: &buf,
		})

		defer Shutdown()

		defer func() {
			if r := recover(); r != nil {
				t.Error("\tGiving a nil writer to output should not panic.", failed)
			}
			t.Log("\tGiving a nil writer to output should not panic.", succeed)
		}()

		// Should not panic if writer is nil
		output(nil, "Asdf %d", 2)
	}
}


log/log.go
/**
* Copyright 2016 Comcast Cable Communications Management, LLC
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
* http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
 */

package log

import (
	"fmt"
	"io"
	"os"
	"path"
	"runtime"
	"sync"
	"sync/atomic"
	"time"
)

// Date and time layout for each trace line.
const (
	layout        = "2006/01/02 15:04:05.000000000"
	emptyMessage  = "**** LOG ERROR: MESSAGE IS EMPTY - PLEASE REPORT ****\n"
	LoggingWasOff = "**** LOG WARNING: LOGGING WAS OFF - PLEASE REPORT ****\n"
)

// Formatter provide support for special formatting.
type Formatter interface {
	Format() string
}

// line is passed to the safe write goroutine
// as the string to write to the device.
type line struct {
	w io.Writer
	b []byte
}

// logger maintains internal state for our logger.
type logger struct {
	dest   map[int8]io.Writer
	destMu sync.RWMutex

	mu           sync.Mutex
	wg           sync.WaitGroup
	write        chan line
	exit         chan struct{}
	stallTimeout time.Duration
	enqueTimer   *time.Timer
	bulkTimer    *time.Timer
	bulkLines    map[io.Writer][]byte

	shutdown      bool
	loggingOff    bool
	pendingWrites int32
	prefix        string
	test          int32
}

// logger maintains a pointer to the single logger.
var l = logger{
	enqueTimer: time.NewTimer(time.Hour),
	bulkTimer:  time.NewTimer(time.Hour),
	bulkLines:  make(map[io.Writer][]byte, 2),
	prefix:     "PREFIX",
}

var bulkLogPeriod = int64(time.Second) // For production, we will use 1 sec, but can change for testing.

// SetBulkLogPeriod sets the private value for the bulk log period.
func SetBulkLogPeriod(p time.Duration) {
	atomic.StoreInt64(&bulkLogPeriod, int64(p))
}

// GetBulkLogPeriod retrieves the private value for the bulk log period.
func GetBulkLogPeriod() time.Duration {
	return time.Duration(atomic.LoadInt64(&bulkLogPeriod))
}

// SetStallTimeout sets the stall timeout value.
func SetStallTimeout(t time.Duration) {
	l.mu.Lock()
	l.stallTimeout = t
	l.mu.Unlock()
}

// Init initializes the logging system for use. It can be called
// multiple times to reset the destination.
func Init(prefix string, bufferSize int, dws ...DevWriter) {
	l.mu.Lock()
	defer l.mu.Unlock()

	if l.write != nil {
		// We need to Unlock the mutex before
		// calling Shutdown and get back in.
		l.mu.Unlock()
		{
			// Shutdown the log with the current configuration.
			Shutdown()
		}
		l.mu.Lock()
	}

	// Set user defined values.
	l.prefix = prefix
	l.write = make(chan line, bufferSize)
	l.exit = make(chan struct{})
	l.stallTimeout = 250 * time.Millisecond

	l.destMu.Lock()
	{
		// Create and init the map of devices.
		l.dest = map[int8]io.Writer{
			DevError:   os.Stderr,
			DevPanic:   os.Stderr,
			DevWarning: os.Stderr,

			DevStart:  os.Stdout,
			DevTrace:  os.Stdout,
			DevQuery:  os.Stdout,
			DevData:   os.Stdout,
			DevSplunk: os.Stdout,
		}
	}
	l.destMu.Unlock()

	// If a device is provided, update the writer.
	if dws != nil {
		for _, dw := range dws {
			// Were we asked to update all the devices.
			if dw.Device == DevAll {
				Dev.All(dw.Writer)
				continue
			}

			l.destMu.Lock()
			{
				// Just update the single device.
				l.dest[dw.Device] = dw.Writer
			}
			l.destMu.Unlock()
		}
	}

	// Set the flags.
	l.loggingOff = false
	l.shutdown = false

	// Create the safe writer goroutine to prevent the log
	// from causing the host application to block on log calls.
	l.wg.Add(1)
	go safeWrite()
}

// InitTest configures the logger for testing purposes.
func InitTest(prefix string, bufferSize int, dws ...DevWriter) {
	SetBulkLogPeriod(50 * time.Millisecond)
	Init(prefix, bufferSize, dws...)
	atomic.StoreInt32(&l.test, 1)
}

// Shutdown will wait until all the pending writes are complete.
func Shutdown() {
	// Sleep for a little bit to allow any possible messages that are about to be enqueued to be placed
	// in the channel.
	time.Sleep(100 * time.Millisecond)
	l.mu.Lock()
	{
		l.shutdown = true
		close(l.write)
		close(l.exit)
		l.wg.Wait()
		l.write = nil
		l.exit = nil

		atomic.StoreInt32(&l.test, 0)
	}
	l.mu.Unlock()
}

// dtFile returns the current time and file for logging.
func dtFile(calldepth int, function string) (dateTime string, file string, funcName string, pid int) {
	// Capture the name of the function logging if
	// a function was not provided.
	if function == "" {
		pc := make([]uintptr, calldepth+1)
		runtime.Callers(calldepth, pc)
		f := runtime.FuncForPC(pc[calldepth-1])
		_, funcName = path.Split(f.Name())
	} else {
		funcName = function
	}

	if atomic.LoadInt32(&l.test) == 1 {
		return time.Date(2009, time.November, 10, 15, 0, 0, 0, time.UTC).UTC().Format(layout), "file.go#512", funcName, 69910
	}

	dateTime = time.Now().UTC().Format(layout)

	_, filePath, line, ok := runtime.Caller(calldepth)
	if !ok {
		return dateTime, "unknown.go#0:", "missing", os.Getpid()
	}
	_, file = path.Split(filePath)

	return dateTime, fmt.Sprintf("%s#%d", file, line), funcName, os.Getpid()
}

// output performs the actual write to the destination device.
func output(w io.Writer, format string, a ...interface{}) {
	if w == nil {
		return
	}
	if format == "" {
		format = emptyMessage
	} else if a != nil {
		format = fmt.Sprintf(format, a...)
	}

	if format[len(format)-1] != '\n' {
		format = format + "\n"
	}

	// Create a slice from the string.
	b := []byte(format)

	l.mu.Lock()
	{
		// We are shutting down. Get out of town.
		if l.shutdown {
			l.mu.Unlock()
			return
		}

		// We have turned logging off. Wait here until the existing
		// buffer has been flushed and then we can start again.
		if l.loggingOff {
			if atomic.LoadInt32(&l.pendingWrites) > 0 {
				l.mu.Unlock()
				return
			}

			l.loggingOff = false
			fmt.Fprintf(w, LoggingWasOff)
		}

		l.enqueTimer.Reset(l.stallTimeout)

		// If we can't perform the write within the wait time, then
		// let's not wait and turn off logging.
		select {
		case l.write <- line{w, b}:
			atomic.AddInt32(&l.pendingWrites, 1)
			l.enqueTimer.Stop()
		case <-l.enqueTimer.C:
			l.loggingOff = true
		}
	}
	l.mu.Unlock()
}

// safeWrite is run as a goroutine. It pulls a message from the
// channel and perform the write.
func safeWrite() {
	l.bulkTimer.Reset(GetBulkLogPeriod())

	flush := func() {
		for k, v := range l.bulkLines {
			go func(k io.Writer, v []byte) {
				if _, err := k.Write(v); err != nil {
					fmt.Fprintf(os.Stderr, "safeWrite ERROR: %s\n", err)
				}
			}(k, v)
			delete(l.bulkLines, k)
		}
	}

exitFor:
	for {
		select {
		case ln := <-l.write:
			if ln.w != nil {
				l.bulkLines[ln.w] = append(l.bulkLines[ln.w], ln.b...)
			}
			atomic.AddInt32(&l.pendingWrites, -1)
		case <-l.bulkTimer.C:
			l.bulkTimer.Reset(GetBulkLogPeriod())
			flush()
		case <-l.exit:
			l.bulkTimer.Stop()
			flush()
			time.Sleep(200 * time.Millisecond) // Need to wait for the flush to perform a write
			break exitFor
		}
	}

	l.wg.Done()
}
